### ISSUE TITLE
Grpc.Tools doesn't support same filename in different folders

### ISSUE THREAD

**Issue by Falco20019** [grpc/grpc#17672]
### What version of gRPC and what language are you using?
Grpc.Tools 1.17.1
 
### What operating system (Linux, Windows, …) and version?
Windows 10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
C# MSBuild
 
### What did you do?
One project with two protos having the same filename but in different folders.

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <Protobuf_StandardImportsPath>$(Protobuf_StandardImportsPath);..\protos</Protobuf_StandardImportsPath>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc.Tools" Version="1.17.1" PrivateAssets="All" />
    <Protobuf Include="..\protos\foo.proto" Link="protos\foo.proto" />
    <Protobuf Include="..\protos\dir1\bar.proto" Link="protos\dir1\bar.proto" />
    <Protobuf Include="..\protos\dir2\bar.proto" Link="protos\dir2\bar.proto" />
  </ItemGroup>
</Project>
```
 
### What did you expect to see?
The code should compile, generating 3 cs files.
 
### What did you see instead?
 ```
MSB3105: The item "obj\Debug\netstandard2.0\Bar.cs" was specified more than once in the "Sources" parameter.  Duplicate items are not supported by the "Sources" parameter.
```
 
### Anything else we should know about your project / environment?
We have an extensive library of protos that get reused by different projects that also share some protos. Therefore a common root path is necessary. The name `config.proto` for example is used in a couple of directories.

Until now, I used the following command to generate the files:
```
%protoc_path%\protoc.exe --proto_path=%protoc_path%\..;..\protos --csharp_out=%output_path% --csharp_opt=base_namespace --grpc_out=%output_path% --plugin=protoc-gen-grpc=%protoc-grpc_path%\grpc_csharp_plugin.exe ..\protos\foo.proto ..\protos\dir1\bar.proto ..\protos\dir2\bar.proto
```

The option `base_namespace` placed the generated files in sub-directories and avoided the problem. This problem will still remains for grpc service files, since there is no option like this for `grpc_opt`. But that problem can't be circumvented so easy right now.

/cc @kkm000

**Comment by jtattermusch** [comment]
I can confirm I'm seeing this problem too.

**Comment by jtattermusch** [comment]
Maybe the right solution would be to implement `base_namespace` option for `grpc_csharp_plugin` too and make Grpc.Tools use that option by default (for both protoc and the plugin) - that would prevent the name clash.

**Comment by Falco20019** [comment]
> Maybe the right solution would be to implement `base_namespace` option for `grpc_csharp_plugin` too and make Grpc.Tools use that option by default (for both protoc and the plugin) - that would prevent the name clash.

Seems like #13724 might get some attention then.

**Comment by jtattermusch** [comment]
Perhaps this has been already discussed, but would this work?
- if .proto file's location within the project is  `foo/bar/some.proto`, can we just set the output directory for that specific .proto file to  `obj/Debug/framework/foo/bar/` ? The C# compiler doesn't really care where is the .cs file located and this would prevent the name collision? At the same time, the output directory is easy to determine as it's just the path of the original .proto file within the file.

**Comment by kkm000** [comment]
Yes, I'm thinking along the same lines. The issue here is some files may be inside the project cone, some outside. Technically possibly with a patch including Windows drive, and so on. I am going to find a way to cover all corner cases. Maybe the same trick with a hash as part of filename that I used for protodep files. It's the easiest, but, unlike the .protodep files, the generated .cs sources may be opened in visual studio--if you look for a symbol definition, for example. I'd rather keep them not very ugly. Also, it cannot be applied blindly anyway, as it is important not to mangle the filename if the user is compiling to generate sources but not compile them (the case we also support, like place all .cs files besides each of the source .protos).

**Comment by jtattermusch** [comment]
- for placing files along the .proto files, we should support the "base_namespace" option I think (which might involve some extra mangling)
- for generating under obj/Debug/ we can use the file's path within the project because the exact path doesn't really matter and it seems it would make stuff much easier
-  For files outside of the project, there is always a "Link" attribute that tells us the "virtual" path of the file within a project, right?

**Comment by kkm000** [comment]
The Link is entirely optional, we cannot rely on it. It just tells VS how to display the file, as if it were at this path under the project node in the tree. For example, for a dotnet console project under Linux there is no point setting it. But I guess we could use it if provided, to treat the file as if it was located in the path relative to the project at the locations indicated by Link.

base_namespace is quite tricky. The big part of the tooling deal with figuring out where to expect the generated file, for dependency checking and to put it into the Compile collection. With base_namespace this becomes impossible, so the only possible source of truth is the protodep file. I need to play a bit with this.

Maybe the best approach is to create the output subdirectory structure as you suggest only for files within the cone (including "virtual" paths from the Link if provided), and change nothing else for other files, at least for starters?

@Falco20019, could you please explain how your .csproj files and .proto files are actually organized with respect to each other? And are you using the Link metadata on .proto files?

**Comment by jtattermusch** [comment]
> The Link is entirely optional, we cannot rely on it. It just tells VS how to display the file, as if it were at this path under the project node in the tree. For example, for a dotnet console project under Linux there is no point setting it. But I guess we could use it if provided, to treat the file as if it was located in the path relative to the project at the locations indicated by Link.
> 
> base_namespace is quite tricky. The big part of the tooling deal with figuring out where to expect the generated file, for dependency checking and to put it into the Compile collection. With base_namespace this becomes impossible, so the only possible source of truth is the protodep file. I need to play a bit with this.
> 
> Maybe the best approach is to create the output subdirectory structure as you suggest only for files within the cone (including "virtual" paths from the Link if provided), and change nothing else for other files, at least for starters?

That sounds good. 
I think requiring "Link" to be set to prevent file name collision for outside-of-project-dir .proto files of the same name is reasonable (otherwise the .proto files would show as "duplicate" in VS project explorer, so not able to build with that setup is not all that surprising and it can be easily worked around by setting the "Link" attr.).

> 
> @Falco20019, could you please explain how your .csproj files and .proto files are actually organized with respect to each other? And are you using the Link metadata on .proto files? Yes, I see the sample repro you've provided, but I'm wondering about your real situation.

**Comment by kkm000** [comment]
@Falco20019, could you please explain how your .csproj files and .proto files are *actually* organized with respect to each other? And are you using the Link metadata on .proto files? Will the above proposed solution work for you? If not, then please explain a little bit about the "framework" that you mentioned. I think I understand that one person/team creates *something*, which is then used by multiple others, and it is the point they are using it the problem happens. I do not understand what is this *something* is: a project on a common file server included by many in their solutions, or a nuget tooling package (i. e. one with props/targets), a Git submodule, etc.

I assume the example you provided as a reproduction for the error, which may or may mot reflect the actual usage pattern. It would be much easier if we understood the real situation and the real problem that you are facing. We could figure out something workable together, I'm sure!

Of course, if nothing else works, you can continue to protoc just as you did pre-1.7.0, with the added (but lesser in this case) help of the package's tooing discovering the required directories for you; read [this section of BUILD-INTEGRATION.md](https://github.com/grpc/grpc/blob/master/src/csharp/BUILD-INTEGRATION.md#alas-this-all-is-nice-but-my-scenario-is-more-complex--or-). But I think it would be better in the end for you and for the community if we could understand the exact problem you are facing, as a fix, if possible, would be beneficial to others, I believe. So let's try to see if there is a possible fix to cover your actual scenario!

**Comment by Falco20019** [comment]
TL;DR: Sorry, was out of office. I am fine with requiring the `Link` attribute to be unique, but would propose to default it to the `%(Identity)`. In fact, we are even enforcing it ourselves for out-of-directory ones. For better understanding, I will still write our use-case up here.

### Complex use-case

The real use-case is a lot more complex and I simplified it for the cause of the ticket. Our real environment looks the following:
- We have a lot of different GRPC microservices, hosting `0..n` GRPC service implementations. Each GRPC microservice is one `csproj` and usually has it's own `git` repository.
- Each microservice may have a couple service-specific protos in a local `protos` folder. This includes for example the configuration file of the service which is also written in `ProtocolBuffers`.
- We have a shared repository of protos that are well-known. This includes the service-protos that are implemented in the GRPC microservices or other well-known protos.
- Our build server will create a NuGET package whenever there are additions on our protos repository. This NuGET package includes a property pointing to the packages included proto files (`$(LIProtoPath)`, see [1]). It also has a `targets` file which adds a `Link` attribute to put the file under `protos/ref/` in the source tree (see [2]).
- Our microservices import a `props` file from a NuGET-stored SDK (`Company.Product.Sdk`) which will add a `PackageReference` the proto-NuGET package, which includes the `props` and `targets` files into the build chain and allows referencing the well-known protos by using `$(LIProtoPath)`-relative paths (see [3]).

### [1] Company.Product.Protos.props
```
<Project>
  <PropertyGroup>
    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
  </PropertyGroup>

  <PropertyGroup>
    <LIProtoPath>$( [System.IO.Path]::GetFullPath($(MSBuildThisFileDirectory)..\protos) )</LIProtoPath>
    <EnableDefaultProtoBufItems>true</EnableDefaultProtoBufItems>
  </PropertyGroup>
</Project>
```

### [2] Company.Product.Protos.targets
```
<?xml version="1.0"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <PropertyGroup>
        <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
    </PropertyGroup>

    <ItemGroup Condition=" '$(Protobuf_ProjectSupported)' == 'true' and '$(Language)' == 'C#' ">
        <ProtoBuf Update="@(ProtoBuf)">
            <Link Condition=" '%(ProtoBuf.Link)' == '' and $([System.String]::Copy('%(Identity)').Contains('$(LIProtoPath)')) ">protos/ref/$([System.String]::Copy('%(Identity)').Replace('$(LIProtoPath)',''))</Link>
        </ProtoBuf>
    </ItemGroup>

    <Target Name="_LiProtos_CoreCompile" BeforeTargets="Protobuf_BeforeCompile">
        <PropertyGroup Condition=" '$(Protobuf_ProjectSupported)' == 'true' and '$(Language)' == 'C#' ">
            <Protobuf_StandardImportsPath>$(Protobuf_StandardImportsPath);$(LIProtoPath)</Protobuf_StandardImportsPath>
        </PropertyGroup>
    </Target>
</Project>
```

### [3] Company.Product.Component.csproj
```
<Project Sdk="Microsoft.NET.Sdk">
    <Import Project="Grpc.props" Sdk="Company.Product.Sdk" />

<!-- // This will be included by Grpc.props together with a lot of other stuff like Copyright, ...
    <ItemGroup>
        <PackageReference Include="Grpc.Tools" Version="1.*" PrivateAssets="All" />
        <PackageReference Include="Company.Product.Protos" Version="1.*" />
    </ItemGroup>
-->

    <ItemGroup>
        <Protobuf Include="protos\db_config.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\update.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\begin_transaction.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\data.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\data_changed.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\delete.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\insert.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\parameterized_query.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\query.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\services\db_data_manipulation.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\services\db_data_manipulation_transactional.proto" Access="Internal" />
    </ItemGroup>

</Project>
```

**Comment by Falco20019** [comment]
TL;DR: Sorry, was out of office. I am fine with requiring the `Link` attribute to be unique, but would propose to default it to the `%(Identity)`. In fact, we are even enforcing it ourselves for out-of-directory ones. For better understanding, I will still write our use-case up here.

### Complex use-case

The real use-case is a lot more complex and I simplified it for the cause of the ticket. Our real environment looks the following:
- We have a lot of different GRPC microservices, hosting `0..n` GRPC service implementations. Each GRPC microservice is one `csproj` and usually has it's own `git` repository.
- Each microservice may have a couple service-specific protos in a local `protos` folder. This includes for example the configuration file of the service which is also written in `ProtocolBuffers`.
- We have a shared repository of protos that are well-known. This includes the service-protos that are implemented in the GRPC microservices or other well-known protos.
- These services can be called from external partners or by our own microservices by referencing the them through a NuGET package.
- Our build server will create this NuGET package whenever there are additions on our protos repository. This NuGET package includes a property pointing to the packages included proto files (`$(LIProtoPath)`, see [1]). It also has a `targets` file which adds a `Link` attribute to put the file under `protos/ref/` in the source tree (see [2]).
- Our microservices import a `props` file from a NuGET-stored SDK (`Company.Product.Sdk`) which will add a `PackageReference` the proto-NuGET package, which includes the `props` and `targets` files into the build chain and allows referencing the well-known protos by using `$(LIProtoPath)`-relative paths (see [3]).
- No one can ensure, that there won't be two protos with the same file name like `$(LIProtoPath)\company\product\db\update.proto` and `$(LIProtoPath)\company\product\device\update.proto`.

### [1] build/Company.Product.Protos.props (part of NuGET Company.Product.Protos)
```
<Project>
  <PropertyGroup>
    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
  </PropertyGroup>

  <PropertyGroup>
    <LIProtoPath>$( [System.IO.Path]::GetFullPath($(MSBuildThisFileDirectory)..\protos) )</LIProtoPath>
    <EnableDefaultProtoBufItems>true</EnableDefaultProtoBufItems>
  </PropertyGroup>
</Project>
```

### [2] build/Company.Product.Protos.targets (part of NuGET Company.Product.Protos)
```
<?xml version="1.0"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <PropertyGroup>
        <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
    </PropertyGroup>

    <ItemGroup Condition=" '$(Protobuf_ProjectSupported)' == 'true' and '$(Language)' == 'C#' ">
        <ProtoBuf Update="@(ProtoBuf)">
            <Link Condition=" '%(ProtoBuf.Link)' == '' and $([System.String]::Copy('%(Identity)').Contains('$(LIProtoPath)')) ">protos/ref/$([System.String]::Copy('%(Identity)').Replace('$(LIProtoPath)',''))</Link>
        </ProtoBuf>
    </ItemGroup>

    <Target Name="_LiProtos_CoreCompile" BeforeTargets="Protobuf_BeforeCompile">
        <PropertyGroup Condition=" '$(Protobuf_ProjectSupported)' == 'true' and '$(Language)' == 'C#' ">
            <Protobuf_StandardImportsPath>$(Protobuf_StandardImportsPath);$(LIProtoPath)</Protobuf_StandardImportsPath>
        </PropertyGroup>
    </Target>
</Project>
```

### [3] Company.Product.Component.DatabaseProvider.csproj
```
<Project Sdk="Microsoft.NET.Sdk">
    <Import Project="Grpc.props" Sdk="Company.Product.Sdk" />

<!-- // This will be included by Grpc.props together with a lot of other stuff like Copyright, ...
    <ItemGroup>
        <PackageReference Include="Grpc.Tools" Version="1.*" PrivateAssets="All" />
        <PackageReference Include="Company.Product.Protos" Version="1.*" />
    </ItemGroup>
-->

    <ItemGroup>
        <Protobuf Include="protos\db_config.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\begin_transaction.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\update.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\delete.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\insert.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\parameterized_query.proto" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\query.proto" Access="Internal" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\services\db_data_manipulation.proto" Access="Internal" GrpcServices="Server" />
        <Protobuf Include="$(LIProtoPath)\company\product\db\services\db_data_manipulation_transactional.proto" Access="Internal" GrpcServices="Server" />
    </ItemGroup>

</Project>
```

`Company.Product.Component.DatabaseProvider` is a microservice, offering two GRPC services for, one for non-transactional (single call) use and one for transactional (bidi-stream) use.

**Comment by kkm000** [comment]
@Falco20019, thanks for the detailed description, I'll be looking into it. Sorry it takes me longer than I planned.

**Comment by Falco20019** [comment]
@kkm000 Any news on this?

**Comment by Falco20019** [comment]
@kkm000 If you need assistance, let me know. I'm not so deep into MSBuild but if it helps speeding this up, I'm willing to spend some time

**Comment by kkm000** [comment]
@Falco20019, so sorry, I managed to slack on this issue again. Being the jack-of-all-trades in a small company is an amazingly messy business! (Please take this not as my attempt at excuse, rather as a statement of fact). I'm hard-planning now a dedicated time to work on gRPC build on the week of 08/19.

**Comment by zhangpengchen** [comment]
As the issue was opened almost a year now, I don't expect it will be fixed soon, I will go for a workdaround

`  <ItemGroup>
    <Protobuf Include="..\protos\dir1\bar.proto" OutputDir="$(Protobuf_OutputPath)dir1" GrpcServices="Server"/>
    <Protobuf Include="..\protos\dir2\bar.proto" OutputDir="$(Protobuf_OutputPath)dir2" GrpcServices="Server"/>
  </ItemGroup>`

**Comment by Falco20019** [comment]
We have to do that too... but it’s annoying if you have dozens of projects and a dozen people doing it manually.

**Comment by godefroi** [comment]
I'm having the same issue, attempting to compile the .proto files for the envoy xDS API. There are no fewer than 4 fault.proto files in various places.

**Comment by Falco20019** [comment]
@kkm000 Any news on that? The problem is growing more and more on us. We are using the workaround from @zhangpengchen and have a knowledge-base article on it. But it's still coming up every couple weeks...

**Comment by wicharypawel** [comment]
I am working on load balancing feature for grpc dotnet, this one take me almost a day because I have to integrate with envoy api which has 25+ duplicating files (I have written small program to count if for me) out of 344 files. I am trying to apply workaround however it is painful. 

@jtattermusch or @Falco20019 could you tell me if this can be pullrequested? Without digging into code it seams simple. Where to start?

**Comment by jtattermusch** [comment]
@wicharypawel I would love to have this fixed but I don't have the cycles. Feel free to help out.

**Comment by wicharypawel** [comment]
What about a solution? 💃 

```
<Protobuf Include="Protos\**\*.proto" ProtoRoot="Protos" OutputDir="$(Protobuf_OutputPath)%(RelativeDir)" />
```

**Comment by Falco20019** [comment]
This is still blocked by https://github.com/grpc/grpc/issues/13724 which I created almost 2.5 years ago.

I got some more knowledge about the build system in the last yet and will look into creating a Pull Request based on @wicharypawel‘s approach later today. I will give it a flag to enable as this is not the default behavior and not backed by the plugin.

**Comment by wicharypawel** [comment]
@Falco20019 if you prepare a sample public repository with your case I can try to do some MSBuild magic to make it happend. 

Something simple, I assume there would be two projects and sample helloworld.proto files. Just to quickly reproduce.

**Comment by Falco20019** [comment]
No problem, I just go the clean way and finally bring `base_namespace` to the `grpc_csharp_plugin`. Just created the PR for it: #22627 :)

I will add another PR for adding support to Grpc.Tools right now.

**Comment by wicharypawel** [comment]
@Falco20019, I have created PR with update to Grpc.Tools documentation, with proposed solution, mainly for those people who are looking for solution today. When your PR will be merged I would ask you to update documentation.

**Comment by Falco20019** [comment]
@wicharypawel Great! This will help users until Grpc.Tool is finally able to support it fully. In #22630, I already updated the documentation. It was a bit of work necessary to get the namespace resolution into MSBuild, but I tried it out locally with different use cases and it seems to work.

Would love to see other also try out the PR and also check the updated documentation as I'm not a native speaker.

**Comment by jtattermusch** [comment]
@wicharypawel  @Falco20019 @kkm000   because Pawel's `%(RelativeDir)` trick seems to work (more discussion on what's the best variable metadata to use in #22627), it seems that we could potentially use this trick by default, by adding it to here: 
https://github.com/grpc/grpc/blob/9c850ff692e5cf3523ff66775857a490a24dad8c/src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets#L30
(that way we could make it the default OutputPath for every Protobuf element if not specified otherwise by the user)

**Comment by Falco20019** [comment]
> @wicharypawel @Falco20019 @kkm000 because Pawel's `%(RelativeDir)` trick seems to work (more discussion on what's the best variable metadata to use in #22627), it seems that we could potentially use this trick by default, by adding it to here:
> 
> https://github.com/grpc/grpc/blob/9c850ff692e5cf3523ff66775857a490a24dad8c/src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets#L30
> 
> 
> (that way we could make it the default OutputPath for every Protobuf element if not specified otherwise by the user)

This does sadly not work if the file is not in the same drive as in my case. See the edit here: https://github.com/grpc/grpc/issues/17672#issuecomment-611329718

**Comment by jtattermusch** [comment]
> @jtattermusch This does sadly not work if the file is not in the same drive as in my case. It would even break it for most of my files with this default. Currently I only have to manually define the duplicates. Then I would have to manually set it for each one. See the edit here: [#17672 (comment)](https://github.com/grpc/grpc/issues/17672#issuecomment-611329718)
> 
> My protos are located on C: in the userprofile as they are distributed through NuGet. My code lies on K: which is an encrypted, virtual drive.
> 
> Main problem here is that the relative path needs to go to the drive root and the result otherwise would include „C:„ with the colon being an forbidden character.
> 
> IF base_namespace would be added as an option, we could use this workaround as default on the location you showed. Because then it would get overwritten if EnableBaseNamespace is set on the item and everything would work even in my code base again. And most users could still just use default with the README only covering the case of different root and manual definition of base_namespace.
> 
> Another dirty option would be to strip out the root part, but this would still result in very long file paths and I already have problems with long paths in other locations and programs on Windows... (i.e. https://winscp.net/forum/viewtopic.php?t=29124)


Even if `%(RelativeDir)` doesn't work for paths on a different drive (= out-of project files), `%(Directory)` or `%(Link)` still might. Can you test that?

**Comment by jtattermusch** [comment]
> > @jtattermusch This does sadly not work if the file is not in the same drive as in my case. It would even break it for most of my files with this default. Currently I only have to manually define the duplicates. Then I would have to manually set it for each one. See the edit here: [#17672 (comment)](https://github.com/grpc/grpc/issues/17672#issuecomment-611329718)
> > My protos are located on C: in the userprofile as they are distributed through NuGet. My code lies on K: which is an encrypted, virtual drive.
> > Main problem here is that the relative path needs to go to the drive root and the result otherwise would include „C:„ with the colon being an forbidden character.
> > IF base_namespace would be added as an option, we could use this workaround as default on the location you showed. Because then it would get overwritten if EnableBaseNamespace is set on the item and everything would work even in my code base again. And most users could still just use default with the README only covering the case of different root and manual definition of base_namespace.
> > Another dirty option would be to strip out the root part, but this would still result in very long file paths and I already have problems with long paths in other locations and programs on Windows... (i.e. https://winscp.net/forum/viewtopic.php?t=29124)
> 
> Even if `%(RelativeDir)` doesn't work for paths on a different drive (= out-of project files), `%(Directory)` or `%(Link)` still might. Can you test that?
> 
> The relevant discussion is here: #22628 (sorry for wrong link earlier).

@Falco20019  I think there has to be a way around this. For .proto files inside the project directory, we seem to already have a working solution (`%(RelativeDir)`) but for all files from outside of the project directory, there must be SOME unique virtual path under which Visual Studio displays them in the project explorer, so I think it should be possible to come up with some rule e.g.:
-  if `%(Link)` is set (should be true for all out-of-project files?), use  `$(Protobuf_OutputPath)%(Link)`
- otherwise use `$(Protobuf_OutputPath)%(RelativeDir)` or something in that sense.

If this doesn't work, I'm sure there's other MSBuild metadata items we could try (they are generally not documented that well so it may be a bit harder to find them).

WDYT?  Can you try with your project structure and post back the results?

**Comment by jtattermusch** [comment]
I've looked at https://github.com/grpc/grpc/pull/22630 and it seems too fragile and complicated for what we want to do.
We should investigate:
- the MSBuild way of solving (as discussed in my previous comment this https://github.com/grpc/grpc/issues/17672#issuecomment-621039153)
- one potential approach would be to use similar trick as the one we used for disambiguating the `.protodep` files (which are also generated by Grpc.Tools).  See here: https://github.com/grpc/grpc/blob/d86acba1daa56ff4867280cb5b4e238d58869ced/src/csharp/Grpc.Tools/DepFileUtil.cs#L166. The idea is that perhaps we could also hash the proto file's full path (regardless what it originally was) and put the generated .cs file under `OutputPath/hash_of_the_full_path/Helloworld.cs`. That way the generated .cs file would have unique path and it would be easy for Grpc.Tools to compute what that path should be.

**Comment by Falco20019** [comment]
@jtattermusch Sounds like a great idea to me and would be a lot less fragile. And it would also solve the long path problem. So I would be fine with investigating that route as default.

I‘m currently out of office for the rest of the week. If no one else is looking into it, I might be able to look into it on the weekend or next week.
